be/src/storage/primary_index.cpp

```diff
@@ -1208,6 +1242,38 @@ Status PrimaryIndex::_upsert_into_persistent_index(uint32_t rssid, uint32_t rowi
     return st;
 }
 
+Status PrimaryIndex::_upsert_into_persistent_index(uint32_t rowset_id, UpdateInfos& info, DeletesMap* deletes,
+                                                   IOStat* stat) {
+    Status st;
+    uint32_t n = info.total_update_rows();
+    std::vector<Slice> keys;
+    std::vector<uint64_t> values;
+    values.reserve(n);
+    std::vector<uint64_t> old_values(n, NullIndexValue);
+    const Slice* vkeys = _build_persistent_keys(info._pk_columns, info._idx_range, &keys);
+    RETURN_IF_ERROR(_build_persistent_values(rowset_id, info, &values));
+    RETURN_IF_ERROR(_persistent_index->upsert(n, vkeys, reinterpret_cast<IndexValue*>(values.data()),
+                                              reinterpret_cast<IndexValue*>(old_values.data()), stat));
```

Comment:
what if there are duplicate keys in vkeys.

Reply:
> what if there are duplicate keys in vkeys.

We will first update these keys in l0 memory table and get the old value from historical l1 files. So if there are duplicate keys(key1) in vkeys, the first key1 will insert into l0 memory and we will mark it as not found in l0 and get old value from l1 files. When we upsert second key1, the first key1 is in l0 memory table and we will mark it as delete.